src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2022 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import data.nat.choose.central
+import algebra.big_operators.fin
+import tactic.field_simp
+
+/-!
+# Catalan numbers
+
+The Catalan numbers (http://oeis.org/A000108) are probably the most ubiquitous sequence of integers
+in mathematics. They enumerate several important objects like binary trees, Dyck paths, and
+triangulations of convex polygons.
+
+## Main definitions
+
+* `catalan n`: the `n`th Catalan number, defined recursively as
+               `catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)`.
+* `catalan' n`: the `n`th Catalan number, defined explicitly as
+               `catalan' n = 1 / (n + 1) * nat.central_binom n`.
+
+## Main results
+
+* `catalan_defs_agree`: The fact that the two definitions both give the same sequence.
+
+## Implementation details
+
+The proof of the equality of the two definitions follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+def catalan' (n : ℕ) :  ℚ :=
+1 / (n + 1) * nat.central_binom n
```

Comment:
I would have expected this to be stated as
```suggestion
def catalan' (n : ℕ) : ℕ :=
nat.central_binom n / (n + 1)
```

Reply:
I agree with the order of division and multiplication. I'm not sure I agree on the `ℚ` vs `ℕ` part. It is not completely trivial to see that `n+1` divides `nat.central_binom n`.

Reply:
If you don't change it to `ℕ` then you can't state that `catalan' n = catalan n` without adding a coercion

Reply:
To differ, I don't really see the point of changing it to `ℕ`.  The fact that `n + 1` divides it is a consequence of the the main theorem `catalan_eq_central_binom_div` that `catalan n = nat.central_binom n / (n + 1)`.  Instead, we could just make `catalan'` private.

Reply:
If it's to be public then it should totally be `\N`. If it's private it can be whatever, of course.